client/pluginmanager/drivermanager/instance.go

```diff
@@ -0,0 +1,443 @@
+package drivermanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/plugins/base"
+	"github.com/hashicorp/nomad/plugins/drivers"
+	"github.com/hashicorp/nomad/plugins/shared/loader"
+	"github.com/hashicorp/nomad/plugins/shared/singleton"
+)
+
+const (
+	// driverFPBackoffBaseline is the baseline time for exponential backoff while
+	// fingerprinting a driver.
+	driverFPBackoffBaseline = 5 * time.Second
+
+	// driverFPBackoffLimit is the limit of the exponential backoff for fingerprinting
+	// a driver.
+	driverFPBackoffLimit = 2 * time.Minute
+)
+
+// instanceManagerConfig configures a driver instance manager
+type instanceManagerConfig struct {
+	// Logger is the logger used by the driver instance manager
+	Logger log.Logger
+
+	// Ctx is used to shutdown the driver instance manager
+	Ctx context.Context
+
+	// Loader is the plugin loader
+	Loader loader.PluginCatalog
+
+	// StoreReattach is used to store a plugins reattach config
+	StoreReattach StorePluginReattachFn
+
+	// FetchReattach is used to retrieve a plugin's reattach config
+	FetchReattach FetchPluginReattachFn
+
+	// PluginConfig is the config passed to the launched plugins
+	PluginConfig *base.ClientAgentConfig
+
+	// ID is the ID of the plugin being managed
+	ID *loader.PluginID
+
+	// updateNodeFromDriver is the callback used to update the node from fingerprinting
+	UpdateNodeFromDriver UpdateNodeDriverInfoFn
+}
+
+// instanceManager is used to manage a single driver plugin
+type instanceManager struct {
+	// logger is the logger used by the driver instance manager
+	logger log.Logger
+
+	// ctx is used to shutdown the driver manager
+	ctx context.Context
+
+	// cancel is used to shutdown management of this driver plugin
+	cancel context.CancelFunc
+
+	// loader is the plugin loader
+	loader loader.PluginCatalog
+
+	// storeReattach is used to store a plugins reattach config
+	storeReattach StorePluginReattachFn
+
+	// fetchReattach is used to retrieve a plugin's reattach config
+	fetchReattach FetchPluginReattachFn
+
+	// pluginConfig is the config passed to the launched plugins
+	pluginConfig *base.ClientAgentConfig
+
+	// id is the ID of the plugin being managed
+	id *loader.PluginID
+
+	// plugin is the plugin instance being managed
+	plugin loader.PluginInstance
+
+	// driver is the driver plugin being managed
+	driver drivers.DriverPlugin
+
+	// pluginLock locks access to the driver and plugin
+	pluginLock sync.Mutex
+
+	// shutdownLock is used to serialize attempts to shutdown
+	shutdownLock sync.Mutex
+
+	// updateNodeFromDriver is the callback used to update the node from fingerprinting
+	updateNodeFromDriver UpdateNodeDriverInfoFn
+
+	// handlers is the map of taskID to handler funcs
+	handlers     map[string][]EventHandler
```

Comment:
Why could there be multiple handlers for a task event? If there's a valid reason let's mention it in a comment as it seems non-intuitive. If there's not let's get rid of the slice.

Reply:
Good catch I thought of this when implementing the Deregister, since it just deletes the entry from the map. Will remove.